pyfr/plugins/tavg.py

```diff
@@ -132,20 +162,85 @@ def _eval_acc_exprs(self, intg):
             exprs.append([npeval(v, subs) for v in self.aexprs])
 
         # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
+        return [np.stack(exs) for exs in exprs]
+
+    def _eval_fun_exprs(self, avars):
 
-    def _eval_fun_exprs(self, intg, accex):
+        # Prepare the substitution dictionary
+        subs = dict(zip(self.anames, avars))
+
+        # Evaluate the function 
+        fexp = [npeval(v, subs) for v in self.fexprs]
+        
+        # Stack the expressions and return
+        return np.stack(fexp)
+
+    def _eval_fun_var(self, dev, accex):
+        dfexpr = []
         exprs = []
+        eps = self.eps
 
         # Iterate over each element type our averaging region
         for avals in accex:
-            # Prepare the substitution dictionary
-            subs = dict(zip(self.anames, avals.swapaxes(0, 1)))
+            df = []
+
+            # Evaluate the function
+            fx = self._eval_fun_exprs(avals)
+            exprs.append(fx)
+
+            for idx, avar in enumerate(self.anames):
+                # Calculate step size 
+                h = np.zeros_like(avals) 
+                h[idx] = np.sqrt(eps) * avals[idx]
```

Comment:
Are we sure that `avals` is always positive?  `np.maximum` can probably clean this up a bit.